File: Day 1/Day1.java
import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

public class Day1 {
    public static void main(String[] args) {
        List<List<Integer>> input = readInput("./Day 1/input.txt");
        if (input.isEmpty()) {
            return;
        }

        System.out.println("distance: " + calculateDistance(input.get(0), input.get(1)));
        System.out.println("similiarity: " + calculateSimiliarity(input.get(0), input.get(1)));
    }

    private static List<List<Integer>> readInput(String path) {
        try {
            File file = new File(path);
            Scanner reader = new Scanner(file);

            List<Integer> listA = new ArrayList<>();
            List<Integer> listB = new ArrayList<>();
            String[] data = null;
            while (reader.hasNextLine()) {
                data = reader.nextLine().split(" ");
                listA.add(Integer.valueOf(data[0]));
                listB.add(Integer.valueOf(data[data.length - 1]));
            }
            reader.close();

            return List.of(listA, listB);

        } catch (Exception e) {
            System.err.println("Error when read input file: ");
            e.printStackTrace();

            return List.of();
        }
    }

    private static Integer calculateDistance(List<Integer> listA, List<Integer> listB) {
        listA.sort(Comparable::compareTo);
        listB.sort(Comparable::compareTo);

        Integer result = 0;
        for (int i = 0; i < listA.size(); i++) {
            result += Math.abs(listA.get(i) - listB.get(i));
        }

        return result;
    }

    private static Integer calculateSimiliarity(List<Integer> listA, List<Integer> listB) {
        HashMap<Integer, Integer> appearanceMap = new HashMap<>();
        for (Integer data : listA) {
            appearanceMap.put(data, 0);
        }

        for (Integer data : listB) {
            if (!appearanceMap.containsKey(data)) {
                continue;
            }
            appearanceMap.put(data, appearanceMap.get(data) + 1);
        }

        Integer result = 0;
        for (Map.Entry<Integer, Integer> entry : appearanceMap.entrySet()) {
            result += entry.getKey() * entry.getValue();
        }

        return result;
    }
}

File: Day 1/day1.py
def read_input(path: str):
    file = open(path, "r")
    list_a = []
    list_b = []
    for line in file:
        data = line.split()
        list_a.append(int(data[0]))
        list_b.append(int(data[-1]))

    return list_a, list_b


def calculate_distance(list_a, list_b):
    sorted_a = sorted(list_a)
    sorted_b = sorted(list_b)

    result = 0
    for i in range(len(sorted_a)):
        result += abs(sorted_a[i] - sorted_b[i])

    return result


def calculate_similiarity(list_a, list_b):
    appearence_dict = {}
    for data in list_a:
        appearence_dict[data] = 0

    for data in list_b:
        if data not in appearence_dict.keys():
            continue

        appearence_dict[data] += 1

    result = 0
    for key, val in appearence_dict.items():
        result += (key * val)

    return result


input = read_input("./input.txt")
print("distance:\t ", calculate_distance(*input))
print("similiarity:\t ", calculate_similiarity(*input))

